test/runtests.py

```diff
@@ -98,126 +130,302 @@ def log(self, filename=None, folder=None, fail=False):
             r.pass_count += 1
             self.pass_count += 1
 
-test_result = TestResult()
-
-
-def show_failed(filename, output, exit_code, expected_output):
-    print("\nFailed -> {}".format(filename))
-    if expected_output == None:
-        print("\nOutput:")
-        print("----------------------------")
-        print(output)
+# test variants:
+#   interpreted: -maxInterpretCount:1 -maxSimpleJitRunCount:1 -bgjit-
+#   dynapogo: -forceNative -off:simpleJit -bgJitDelay:0
+class TestVariant(object):
+    MSG_PRINT = 0
+    MSG_TEST_RESULT = 1
+    _empty_set = set()
+
+    def __init__(self, name, compile_flags=[]):
+        self.name = name
+        self.compile_flags = \
+            ['-WERExceptionSupport', '-ExtendedErrorStackForTestHost'] + compile_flags
+        self.tags = tags.copy()
+        self.not_tags = not_tags.union(
+            ['{}_{}'.format(x, name) for x in 'fails','excludes'])
+
+        self.msg_queue = Manager().Queue() # messages from multi processes
+        self.test_result = TestResult()
+
+    # check if this test variant should run a given test
+    def _should_test(self, test):
+        tags = test.get('tags')
+        tags = set(tags.split(',')) if tags else self._empty_set
+        if not tags.isdisjoint(self.not_tags):
+            return False
+        if self.tags and not self.tags.issubset(tags):
+            return False
+        return True
+
+    # queue a test result from multi-process runs
+    def _log_result(self, filename, fail=False):
+        self.msg_queue.put((self.MSG_TEST_RESULT, filename, fail))
```

Comment:
As I understand from reading multiprocessing docs, Manager().Queue() is "thread and process safe". Each time I sent a single complete message -- a test pass/fail result, or a piece of output. So no lock is needed.